python/ray/tests/test_client_multi.py

```diff
@@ -144,5 +153,32 @@ def double(self):
         ray.get(o1)
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="PSUtil does not work the same on windows.")
+@pytest.mark.parametrize(
+    "call_ray_start",
+    ["ray start --head --ray-client-server-port 25001 --port 0"],
+    indirect=True)
+def test_multi_cli_threading(call_ray_start):
+    import threading
+    b = threading.Barrier(2)
+
+    def get(idx):
+        cli = ray.init("ray://localhost:25001", allow_multiple=True)
+        with cli:
+            a = ray.put(20)
+            b.wait()
+            assert 20 == ray.get(a)
+            b.wait()
+
+    t1 = threading.Thread(target=get, args=(0, ))
+    t2 = threading.Thread(target=get, args=(1, ))
+    t1.start()
+    t2.start()
+    t1.join()
+    t2.join()
```

Comment:
I think a failure in the thread won't actually fail the test. Can you maybe have the thread push to a shared queue on success and assert on the main thread that two entries are in the queue?